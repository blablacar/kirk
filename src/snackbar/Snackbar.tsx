import React, { PureComponent } from 'react'
import { createPortal } from 'react-dom'
import TransitionGroup from 'react-transition-group/TransitionGroup'
import cc from 'classcat'
import { canUseDOM } from 'exenv'

import { color } from '_utils/branding'
import Button, { ButtonStatus } from 'button'
import CrossIcon from 'icon/crossIcon'
import Text, { TextDisplayType } from 'text'
import CustomTransition, { AnimationType } from 'transitions'

export interface SnackbarProps {
  readonly close: () => void
  readonly isOpen: boolean
  readonly className?: string
  readonly extraClassName?: string
  readonly children: JSX.Element
}

class Snackbar extends PureComponent<SnackbarProps> {
  private portalNode: HTMLElement
  static defaultProps: Partial<SnackbarProps> = {
    isOpen: false,
  }

  constructor(props: SnackbarProps) {
    super(props)
    if (canUseDOM) {
      this.portalNode = document.createElement('div')
      document.body.appendChild(this.portalNode)
    }
  }

  componentWillUnmount() {
    if (this.portalNode) {
      document.body.removeChild(this.portalNode)
      this.portalNode = null
    }
  }

  render() {
    const modalElement = (
      <TransitionGroup
        component="div"
        className={cc(['transition-wrapper kirk-snackbar-container', this.props.className])}
      >
        {this.props.isOpen && (
          <CustomTransition animationName={AnimationType.SLIDE_UP}>
            <div className={cc(['kirk-snackbar', this.props.extraClassName])}>
              <Text role="alert" display={TextDisplayType.TITLE} className="kirk-snackbar-content">
                {this.props.children}
              </Text>
              <Button
                status={ButtonStatus.UNSTYLED}
                className="kirk-snackbar-cross"
                onClick={this.props.close}
              >
                <CrossIcon iconColor={color.white} size="18px" />
              </Button>
            </div>
          </CustomTransition>
        )}
      </TransitionGroup>
    )

    if (!this.portalNode) {
      return modalElement
    }

    return createPortal(modalElement, this.portalNode)
  }
}

export default Snackbar
