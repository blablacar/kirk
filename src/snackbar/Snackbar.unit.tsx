import React from 'react'
import exenv from 'exenv'
import renderer from 'react-test-renderer'
import { shallow, mount } from 'enzyme'

import Button from 'button'
import Snackbar from './Snackbar'
import StyledSnackbar from './index'

const close = jest.fn()

describe('Snackbar', () => {
  it('should not have changed', () => {
    exenv.canUseDOM = false
    const snackbarServerSide = renderer
      .create(
        <StyledSnackbar isOpen close={close}>
          Oups
        </StyledSnackbar>,
      )
      .toJSON()
    exenv.canUseDOM = true
    expect(snackbarServerSide).toMatchSnapshot()
  })

  it('should show the snackbar when isOpen is true', () => {
    const wrapper = shallow(<Snackbar isOpen close={close} />)
    expect(wrapper.find('.kirk-snackbar').exists()).toBe(true)
  })

  it('should not show the snackbar when isOpen is false', () => {
    const wrapper = shallow(<Snackbar isOpen={false} close={close} />)
    expect(wrapper.find('.kirk-snackbar').exists()).toBe(false)
  })

  it('should call close method when click on cross button', () => {
    const wrapper = shallow(<Snackbar isOpen close={close} />)
    wrapper.find(Button).simulate('click')
    expect(close).toHaveBeenCalled()
  })

  it('should render the same layout on client and server side', () => {
    exenv.canUseDOM = false
    const serverSide = mount(<Snackbar isOpen close={close} />)
    exenv.canUseDOM = true
    const clientSide = mount(<Snackbar isOpen close={close} />)

    expect(serverSide.html()).toEqual(clientSide.html())
  })
})
