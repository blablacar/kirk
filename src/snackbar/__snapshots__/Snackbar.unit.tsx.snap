// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snackbar should not have changed 1`] = `
.c1 {
  margin: 0;
  font-weight: 400;
}

.c1.kirk-text-button {
  color: #054752;
  font-size: 16px;
  line-height: 20px;
}

.c1.kirk-text-body,
.c1.kirk-text-bodyStrong {
  color: #708C91;
  font-size: 16px;
  line-height: 20px;
}

.c1.kirk-text-caption {
  color: #708C91;
  font-size: 13px;
  line-height: 16px;
}

.c1.kirk-text-title,
.c1.kirk-text-titleStrong {
  color: #054752;
  font-size: 18px;
  line-height: 20px;
}

.c1.kirk-text-display1 {
  color: #054752;
  font-size: 82px;
  line-height: 82px;
}

.c1.kirk-text-display2 {
  color: #054752;
  font-size: 30px;
  line-height: 1.06;
}

.c1.kirk-text-subheader,
.c1.kirk-text-subheaderStrong {
  color: #054752;
  font-size: 22px;
  line-height: 24px;
}

.c1.kirk-text-title,
.c1.kirk-text-titleStrong {
  color: #054752;
  font-size: 18px;
  line-height: 20px;
}

.c1.kirk-text-bodyStrong,
.c1.kirk-text-subheaderStrong,
.c1.kirk-text-titleStrong {
  font-weight: 500;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .kirk-snackbar {
  position: fixed;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  width: 100%;
  max-width: 614px;
  margin: 0 auto;
  bottom: 0;
  overflow: hidden;
  background-color: #F53F5B;
  padding: 16px 24px;
  z-index: 1000;
}

.c0 .kirk-snackbar-content {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin: 0;
  color: #FFF;
}

.c0 .kirk-snackbar-cross {
  padding: 0 0 0 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  algin-items: flex-start;
}

<div
  className="transition-wrapper kirk-snackbar-container c0"
>
  <div
    className="kirk-snackbar slide-up slide-up-entered"
  >
    <span
      className="kirk-text kirk-text-title kirk-snackbar-content c1"
      role="alert"
    >
      Oups
    </span>
    <button
      className="kirk-button kirk-button-unstyled kirk-snackbar-cross"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="kirk-icon"
        fill="#FFF"
        height="18px"
        viewBox="0 0 24 24"
        width="18px"
        xmlns="http://www.w3.org/2000/svg"
      >
        
        <path
          d="M19 5L5 19M19 19L5 5"
          fill="none"
          stroke="#FFF"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeMiterlimit="10"
          strokeWidth="2"
        />
      </svg>
    </button>
  </div>
</div>
`;
